From c18691a35c1d560ba0215a60a4273ff8e5794237 Mon Sep 17 00:00:00 2001
From: Yann Gautier <yann.gautier@foss.st.com>
Date: Wed, 30 Oct 2024 11:09:15 +0100
Subject: [PATCH 739/954] refactor(stm32mp2): align
 bl31_plat_get_next_image_ep_info

This function has been reworked upstream. To ease future updates, align
the code of bl31_plat_get_next_image_ep_info() function with upstream.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git a7f0b9886c87c900210329458e134cc26a54280c refactor(stm32mp2): align bl31_plat_get_next_image_ep_info]
Change-Id: If15d9813afd3a81f01a3c1e7eb01a1647d6f6c06
Signed-off-by: Yann Gautier <yann.gautier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/415158
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 plat/st/stm32mp2/bl31_plat_setup.c | 27 ++++++++++++++++++++++-----
 1 file changed, 22 insertions(+), 5 deletions(-)

diff --git a/plat/st/stm32mp2/bl31_plat_setup.c b/plat/st/stm32mp2/bl31_plat_setup.c
index b048cab96c..184b216d59 100644
--- a/plat/st/stm32mp2/bl31_plat_setup.c
+++ b/plat/st/stm32mp2/bl31_plat_setup.c
@@ -178,12 +178,29 @@ void bl31_plat_runtime_setup(void)
 
 entry_point_info_t *bl31_plat_get_next_image_ep_info(unsigned int type)
 {
-	if (type == NON_SECURE)
-		return &bl33_image_ep_info;
-	if (type == SECURE)
-		return &bl32_image_ep_info;
+	entry_point_info_t *next_image_info = NULL;
 
-	return NULL;
+	assert(sec_state_is_valid(type));
+
+	switch (type) {
+	case NON_SECURE:
+		next_image_info = &bl33_image_ep_info;
+		break;
+
+	case SECURE:
+		next_image_info = &bl32_image_ep_info;
+		break;
+
+	default:
+		break;
+	}
+
+	/* None of the next images on ST platforms can have 0x0 as the entrypoint */
+	if ((next_image_info == NULL) || (next_image_info->pc == 0UL)) {
+		return NULL;
+	}
+
+	return next_image_info;
 }
 
 #if STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER
-- 
2.53.0

